### PR TITLE
[RFC] vim-patch:7.4.1037,fa73534

### PR DESCRIPTION
**Pull request by jamessan**
#### vim-patch:7.4.1037

Problem:    Using "q!" when there is a modified hidden buffer does not unload
            the current buffer, resulting in the need to abandon it again.
Solution:   When using "q!" unload the current buffer when needed. (Yasuhiro
            Matsumoto, Hirohito Higashi)

https://github.com/vim/vim/commit/027387f70c671f62e3e08e0bdd09ec05b0232735
#### vim-patch:fa73534

Updated runtime files.

https://github.com/vim/vim/commit/fa7353428f705f7a13465a1943dddeede4083023

Missing in runtime/doc: if_tcl.txt, tags, todo.txt, version5.txt.  Some
other changes related to binary nrformats were already merged.

### CODE REVIEW

**Comment by justinmk** [comment]
quickbuild failure on `functionaltest-freebsd-64-release-539`:

```
test/functional/legacy/031_close_commands_spec.lua @ 19
Failure message: test/functional/legacy/031_close_commands_spec.lua:107: Expected a different error.
Caught:
(no error)
Expected:
(error)
stack traceback:
```

QB results should be valid since they were fixed on master earlier this week.

**Comment by jamessan** [comment]
Yeah, that implies that nvim didn't exit when it should have ...

**Comment by jamessan** [comment]
Hmm, there was actually a segfault in the Debug build.

**Comment by jamessan** [comment]
Ok, so I finally reproduced the crash/segfault by changing

    assert.has_error(function()
      execute('call append(line("$"), "Oh, Not finished yet.")')
    end)

to

    assert.has_error(function()
      expect('testtext 1')
    end)

The backtrace is
````
#0  0x00007f6c1485037a in luv_check_callback (L=0x40f1c378, data=0x0, id=1, index=2)
    at src/lhandle.c:51
#1  0x00007f6c14843f5b in luv_read_start (L=0x40f1c378) at src/stream.c:121
#2  0x0000000000423c67 in lj_BC_FUNCC ()
#3  0x000000000040ed68 in lua_pcall (L=0x40f1c378, nargs=<optimized out>,
    nresults=<optimized out>, errfunc=<optimized out>) at lj_api.c:1052
#4  0x00000000004050c9 in docall (L=0x40f1c378, narg=7, clear=0) at luajit.c:122
#5  0x0000000000404844 in handle_script (L=<optimized out>, argv=<optimized out>,
    n=<optimized out>) at luajit.c:289
#6  pmain (L=0x40f1c378) at luajit.c:538
#7  0x0000000000423c67 in lj_BC_FUNCC ()
#8  0x000000000040ee68 in lua_cpcall (L=0x40f1c378, func=0xffffd8f0, ud=0xfffffff7)
    at lj_api.c:1074
#9  0x0000000000403fc7 in main (argc=13, argv=0x7fff269ae948) at luajit.c:566
````

The intention of that test is to verify that nvim has exited.  In the legacy test, the code just wouldn't be executed (because Vim had exited) so the subsequent command appending to the `.ok` file never happens.

In our case, we're trying to drive nvim through luajit and the disappearing act isn't taken very well.  Is there a better way to perform this test, which is explicitly about the interaction of modified buffers and trying to exit Vim?

File: runtime/doc/editing.txt [hunk 1]
```diff
@@ -1042,10 +1042,10 @@ The names can be in upper- or lowercase.
 			the last file in the argument list has not been
 			edited.  See |:confirm| and 'confirm'.
 
-:q[uit]!		Quit without writing, also when currently visible
-			buffers have changes.  Does not exit when this is the
-			last window and there is a changed hidden buffer.
-			In this case, the first changed hidden buffer becomes
+:q[uit]!		Quit without writing, also when currentl buffer has
```
Comment (justinmk) [inline runtime/doc/editing.txt]: typo: currently

File: test/functional/legacy/031_close_commands_spec.lua [hunk 2]
```diff
@@ -84,6 +84,29 @@ describe('Commands that close windows and/or buffers', function()
...
+    execute('q!')
+    expect('testtext 3')
+    execute('q!')
+    feed('<CR>')
+    expect('testtext 1')
+    execute('q!')
+    -- Now nvim has exited
+    assert.has_error(function()
+      execute('call append(line("$"), "Oh, Not finished yet.")')
+    end)
```
Comment (justinmk) [inline test/functional/legacy/031_close_commands_spec.lua]: To workaround the segfault you could put the code in a single `source([[...]])` call, and that should be equivalent to the old test?

**Comment by justinmk** [comment]
That seems like a bug in our test `helpers.lua` or in the lua-client, or in libmpack.

File: runtime/doc/editing.txt (hunk 1 above)
Comment (jamessan) [inline runtime/doc/editing.txt]: That's fixed in a runtime patch, which I should be able to pull in now.

File: test/functional/legacy/031_close_commands_spec.lua (hunk 2 above)
Comment (jamessan) [inline test/functional/legacy/031_close_commands_spec.lua]: Thanks for the suggestion.  That works.

**Comment by jamessan** [comment]
Do you want me to file a separate issue (although, not sure which repo would be right) for the segfault?

**Comment by justinmk** [comment]
QB: publish-functional-test-report  failed but run-functional-tests succeeded... I _think_ that means we're good?

> Do you want me to file a separate issue (although, not sure which repo would be right) for the segfault?

The segfault looks like a bug https://github.com/luvit/luv .  `luv_check_callback` deferences `data` [here](https://github.com/luvit/luv/blob/29183bd28450e8e0750a36d3c49182a7e650a866/src/lhandle.c#L51). Before it reaches that line it is checked by `luv_check_stream` but that check doesn't check `.data`, unlike other `luv_check_*` functions. I'll go ahead and create a bug report while this is in my head.

**Comment by KillTheMule** [comment]
> QB: publish-functional-test-report failed but run-functional-tests succeeded... I think that means we're good?

If you look at the step `build-and-run-tests-parameterized?buildType=Release` there is an uncaught error with an invalid msgpack-string in  `ttest/functional/terminal/tui_spec.lua:159: in function <test/functional/terminal/tui_spec.lua:151>`.  Doesn't look good to me :(

**Comment by jamessan** [comment]
I think I've seen that `invalid msgpack-string` message in other test runs too.  I'm not sure if that's a flaky test or there's something else going on here.  This PR shouldn't have changed the behavior of that test.  It looks like the same failures that have been seen during some of the QB runs in #4563 (it was the large bracketed paste test).

An odd part is that the `WARN` lines for the unexpected error show up in the log before the log even says it's running the tui_spec test.

**Comment by jszakmeister** [comment]
> QB: publish-functional-test-report failed but run-functional-tests succeeded... I think that means we're good?

No.  Failed test do not cause run-functional-tests to fail.  The TAP plan is checked in the publish step.  It's a bit goofy, but that's the way it works.

**Comment by jszakmeister** [comment]
> No. Failed test do not cause run-functional-tests to fail. The TAP plan is checked in the publish step. It's a bit goofy, but that's the way it works

I changed the names of the steps a little to 'check-and-publish-...'.  Hopefully that makes it a little more clear.

**Comment by jamessan** [comment]
In fact, now #4737 is failing QB tests with the same problem.  What's up with this test on the OSX builder?

**Comment by KillTheMule** [comment]
I've seen it on freebsd as well, and fractions of it on some travis builds.

**Comment by KillTheMule** [comment]
Give https://github.com/neovim/neovim/pull/4563/files#diff-8a087fe7dae801e7d38a9d0af6746a8eR157 a shot, it fixes everything for me.

**Comment by justinmk** [comment]
Thanks everyone. We powered through a lot of robustness issues this week!

**Comment by creationix** [comment]
@justinmk, did you ever find out what was causing the data field to be NULL in luv?  I'm hitting cases in our app where it's triggering my new guard and have no idea what's the root cause of the NULL data property.

**Comment by justinmk** [comment]
@creationix I didn't look for the root cause. Our [use case](https://github.com/neovim/neovim/pull/4743#issuecomment-218944511) is one where we expect a failure (but not a segfault).

**Comment by creationix** [comment]
I figured out our root cause and my patch was actually making things worse.  It should be better now. (the patch I added for this issue was causing it to throw an error if you try to call :is_closing on something that's already closed)

**Pull request merged**